Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |


- In the context of this problem, what is a false positive?
- In the context of this problem, what is a false negative?
- How would you describe this model?


What is the positive and negative case?
- In the context of this problem, what is a false positive?

false negative: predicts that is a cat and is not  /  predicts that is a dog and is not.
- In the context of this problem, what is a false positive?  

false positive: predicts that it is not cat and is  /  predicts that it is not dog and is

How would you describe this model?
precicision

In [1]:
# cat = positive class
# dog = negative class


# False positive: We predicted cat, but it is actually a dog
# False negative: We predicted dog, but it is actually a cat

tp = 34
tn = 46
fp = 7
fn = 13

In [2]:
print("Cat-classifier (where 'cat' is the positive prediction)")
print("True Positives", tp)
print("False Positives", fp)
print("False Negatives", fn)
print("True Negatives", tn)

print("-------------")

accuracy = (tp + tn) / (tp + tn + fp + fn)
recall = tp / (tp + fn)
precision = tp / (tp + fp)

print("Accuracy is", accuracy)
print("Recall is", round(recall,2))
print("Precision is", round(precision,2))

Cat-classifier (where 'cat' is the positive prediction)
True Positives 34
False Positives 7
False Negatives 13
True Negatives 46
-------------
Accuracy is 0.8
Recall is 0.72
Precision is 0.83


### 3) You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [3]:
import numpy as np
import seaborn as sns
import scipy.stats as stats
import pandas as pd
import matplotlib.pyplot as plt
import env
from pydataset import data
import scipy
import os
from sklearn.model_selection import train_test_split

In [4]:
c3 = pd.read_csv('c3.csv')

In [5]:
c3

actual     model1     model2     model3
0    No Defect  No Defect     Defect  No Defect
1    No Defect  No Defect     Defect     Defect
2    No Defect  No Defect     Defect  No Defect
3    No Defect     Defect     Defect     Defect
4    No Defect  No Defect     Defect  No Defect
..         ...        ...        ...        ...
195  No Defect  No Defect     Defect     Defect
196     Defect     Defect  No Defect  No Defect
197  No Defect  No Defect  No Defect  No Defect
198  No Defect  No Defect     Defect     Defect
199  No Defect  No Defect  No Defect     Defect

[200 rows x 4 columns]

In [6]:
c3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   actual  200 non-null    object
 1   model1  200 non-null    object
 2   model2  200 non-null    object
 3   model3  200 non-null    object
dtypes: object(4)
memory usage: 6.4+ KB


In [7]:
c3.describe()

actual     model1     model2     model3
count         200        200        200        200
unique          2          2          2          2
top     No Defect  No Defect  No Defect  No Defect
freq          184        190        110        101

In [8]:
pd.crosstab(c3.model1, c3.actual)

actual     Defect  No Defect
model1                      
Defect          8          2
No Defect       8        182

KeyError: 'No Defect'

In [9]:
pd.crosstab(c3.model2, c3.actual)

actual     Defect  No Defect
model2                      
Defect          9         81
No Defect       7        103

In [10]:
pd.crosstab(c3.model3, c3.actual)

actual     Defect  No Defect
model3                      
Defect         13         86
No Defect       3         98

In [11]:
c3.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [12]:
c3['baseline'] = 'No Defect'

In [13]:
(c3.actual == c3.model1).mean()

0.95

In [14]:
# baseline accuracy
(c3.actual == c3.baseline).mean()

0.92

- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

### Recall, model 3 


- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

### Reacall, model 1

### 4) You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

- Several models have already been developed with the data, and you can find their results here.

- Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

    a) In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?
    
    b) Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?
    
    c) Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

In [42]:
df = pd.read_csv('gives_you_paws.csv')

In [43]:
df

actual model1 model2 model3 model4
0       cat    cat    dog    cat    dog
1       dog    dog    cat    cat    dog
2       dog    cat    cat    cat    dog
3       dog    dog    dog    cat    dog
4       cat    cat    cat    dog    dog
...     ...    ...    ...    ...    ...
4995    dog    dog    dog    dog    dog
4996    dog    dog    cat    cat    dog
4997    dog    cat    cat    dog    dog
4998    cat    cat    cat    cat    dog
4999    dog    dog    dog    dog    dog

[5000 rows x 5 columns]

In [46]:
### teachers code
subset = df[df.actual == 'Defect']
subset

# PR team really wants to minimize the False positives - meaning choose model with highest precision:

# choose subset of model1 where we only select 'positive predictions'

# calculate precision
model_precision = (subset.actual == subset.model1).mean()

print("Model 1")
print(f"Model precision: {model_precision:.2%}")
# choose subset for model3 where we only select 'positive predictions'
subset = df[df.model3 == 'Defect']

# choose subset for model2 where we only select 'positive predictions'
subset = df[df.model2 == 'Defect']

# calculate precision
model_precision = (subset.actual == subset.model2).mean()

print("Model 2")
print(f"Model precision: {model_precision:.2%}")


# calculate precision
model_precision = (subset.actual == subset.model3).mean()

print("Model 3")
print(f"Model precision: {model_precision:.2%}")

Model 1
Model precision: nan%


In [18]:
pd.crosstab(df.model1, df.actual)

actual   cat   dog
model1            
cat     1423   640
dog      323  2614

In [19]:
pd.crosstab(df.model2, df.actual)

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

In [20]:
pd.crosstab(df.model3, df.actual)

actual  cat   dog
model3           
cat     893  1599
dog     853  1655

In [21]:
pd.crosstab(df.model4, df.actual)

actual   cat   dog
model4            
cat      603   144
dog     1143  3110

In [22]:
df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [23]:
df['baseline'] = 'dog'

In [24]:
# baseline accuracy
(df.actual == df.baseline).mean()

0.6508

    a) In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?
### 4th model being the best

In [25]:
# model accuracy
(df.model1 == df.baseline).mean()

0.5874

In [26]:
# model accuracy
(df.model2 == df.baseline).mean()

0.3576

In [27]:
# model accuracy
(df.model3 == df.baseline).mean()

0.5016

In [28]:
# model accuracy
(df.model4 == df.baseline).mean()

0.8506

    b) Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?
    
### For phase I model 3 would be the best, giving us a 50% chance.
### For phase II model 1. - lets in highest mistakes so humans can correct

    c) Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?
### Phase I = model 4
### Phase II = model 2

### 5) Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

In [29]:
from sklearn.metrics import classification_report

In [30]:
subset1 = df[df.model1 == 'dog']
subset2 = df[df.model2 == 'dog']
subset3 = df[df.model3 == 'dog']
subset4 = df[df.model4 == 'dog']

In [31]:
subset1.head()

actual model1 model2 model3 model4 baseline
1    dog    dog    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
5    dog    dog    dog    dog    dog      dog
7    cat    dog    cat    cat    dog      dog
8    dog    dog    cat    dog    dog      dog

In [32]:
# Model 1 Precision
((subset1.actual == subset1.model1).mean(),
# Model 2 Precision
(subset2.actual == subset2.model2).mean(),
# Model 3 Precision
(subset3.actual == subset3.model3).mean(),
# Model 4 Precision
(subset4.actual == subset4.model4).mean())

(0.8900238338440586,
 0.8931767337807607,
 0.6598883572567783,
 0.7312485304490948)

In [33]:
pd.crosstab(df.model2, df.actual)

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

Dog team’s Phase-2 model is Model 2 with Precision of 0.893

In [34]:
from sklearn.metrics import classification_report

In [35]:
x = classification_report(df.actual, df.model1,
                          labels = ['cat', 'dog'],
                          output_dict=True)
pd.DataFrame(x).T

precision    recall  f1-score    support
cat            0.689772  0.815006  0.747178  1746.0000
dog            0.890024  0.803319  0.844452  3254.0000
accuracy       0.807400  0.807400  0.807400     0.8074
macro avg      0.789898  0.809162  0.795815  5000.0000
weighted avg   0.820096  0.807400  0.810484  5000.0000

In [50]:
print("Model 1")
pd.DataFrame(classification_report(df.actual, df.model1,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T

Model 1


precision    recall  f1-score    support
cat            0.689772  0.815006  0.747178  1746.0000
dog            0.890024  0.803319  0.844452  3254.0000
accuracy       0.807400  0.807400  0.807400     0.8074
macro avg      0.789898  0.809162  0.795815  5000.0000
weighted avg   0.820096  0.807400  0.810484  5000.0000

In [37]:
print("Model 2")
pd.DataFrame(classification_report(df.actual, df.model2,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T

Model 2


precision    recall  f1-score    support
cat            0.484122  0.890607  0.627269  1746.0000
dog            0.893177  0.490781  0.633479  3254.0000
accuracy       0.630400  0.630400  0.630400     0.6304
macro avg      0.688649  0.690694  0.630374  5000.0000
weighted avg   0.750335  0.630400  0.631310  5000.0000

In [38]:

print("Model 3")
pd.DataFrame(classification_report(df.actual, df.model3,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T

Model 3


precision    recall  f1-score    support
cat            0.358347  0.511455  0.421425  1746.0000
dog            0.659888  0.508605  0.574453  3254.0000
accuracy       0.509600  0.509600  0.509600     0.5096
macro avg      0.509118  0.510030  0.497939  5000.0000
weighted avg   0.554590  0.509600  0.521016  5000.0000

In [39]:
print("Model 4")
pd.DataFrame(classification_report(df.actual, df.model4,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T

Model 4


precision    recall  f1-score    support
cat            0.807229  0.345361  0.483755  1746.0000
dog            0.731249  0.955747  0.828560  3254.0000
accuracy       0.742600  0.742600  0.742600     0.7426
macro avg      0.769239  0.650554  0.656157  5000.0000
weighted avg   0.757781  0.742600  0.708154  5000.0000

Cat team’s Phase-1 model is Model 2 with Recall of 0.89

Cat team’s Phase-2 model is Model 4 with Precision of 0.81